### ISSUE TITLE
Better variable naming

### ISSUE THREAD

**Issue by rrtoledo** [cardano-scaling/alba#38]
## Why
For people not knowing the Alba paper well, the variable names are hard to follow. 

## What
We suggest here to change the variable names to be more descriptive and add a mapping to the paper.

**Comment by rrtoledo** [comment]
Ideas from @curiecrypt:
```
Sp: prover_set
np: set_size
nf: threshold or verifier_threshold (?)
u: proof_size
d: search_depth
q: selection_probability
r: num_of_repetitions
b: bound
s: set_item
s_list: item_list
t: tuple_index
v: attempt_index
H0: item_hash
H1: prefix_hash
H2: final_hash
```

**Comment by rrtoledo** [comment]
Ideas from @tolikzinovyev :
```
np - prover set size
nf - adversary set size
d - search width / number of search trees
q - final selection probability
r - number of tries (just to make it shorter, attempts is also fine)
b - DFS bound
s - element (to be consistent with the paper)
s_list - I'm guessing it's a (partial) tuple? tuple / element sequence
t - tree index
v - try index (just to make it short, attempt is also fine)
H0 - bin hash function
```

**Comment by curiecrypt** [comment]
@rrtoledo, @djetchev, @tolikzinovyev. Here is a list of all suggestions we have. State your choice by putting `x` beside it.
Also, add more lines if you have more suggestions.

**Sp:**
- prover_set `x`
---
**np:** 
- set_size `x`
- prover set size
---
**nf:** 
- threshold `x`
- verifier_threshold 
- adversary set size
---
**u:** 
- proof_size `x`

**d:** 
- search_depth `x`
- search width 
- number of search trees
---
**q:** 
- selection_probability
- final selection probability `x`
---
**r:** 
- num_of_repetitions
- number of tries `x`
- attempts
---
**b:** 
- bound
- DFS bound `x`
---
**s:** 
- set_item
- element `x`
---
**s_list:** 
- item_list
- tuple
- element sequence `x`
---
**t:** 
- tuple_index 
- tree index `x`
---
**v:** 
- attempt_index `x`
- try index
---
**H0:** 
- item_hash
- bin hash function
- element_hash `x`
---
**H1:** 
- prefix_hash `x`
---
**H2:** 
- final_hash `x`

**Comment by rrtoledo** [comment]
@curiecrypt I reordered the variables (to link `t` with `d`, and `v` with `r`), added a few options and voted.
I have a preference for Merkle/tree terminology as I think it makes it easier to reason with (especially with a whiteboard).

I am not sure about `d`, I chose so far `search_width` but I wonder if choosing forest/polytree might be better: we keep the picture in mind, but forests/polytrees are not as "famous" as trees. We also have to agree first on our representation: whether the trees are disjoint (forest) or not (polytree).

**Comment by tolikzinovyev** [comment]
Added my letter `T` to my choices. No matter what we choose, we should make sure the naming is consistent. For example, if `v` is "attempt index", `r` should be "number of attempts", not "number of tries". Or if `q` is "final_selection_probability", `H2` should be "final hash".

**Comment by curiecrypt** [comment]
> Added my letter `T` to my choices. No matter what we choose, we should make sure the naming is consistent. For example, if `v` is "attempt index", `r` should be "number of attempts", not "number of tries". Or if `q` is "final_selection_probability", `H2` should be "final hash".

@tolikzinovyev I agree with you about the consistency. 

Now, I will make a list for the final version of each variable. @rrtoledo , @tolikzinovyev let me know if you have any problems with the final decision.

**Comment by rrtoledo** [comment]
Looks good, just one thing, if we choose `d: search_width`, to be consistent as @tolikzinovyev said, let's rename `t:search_index` (or `t:search_counter`)

**Comment by tolikzinovyev** [comment]
Looks good, I'm just not sure about `nf`. What are the tradeoffs between different options? Maybe I just don't understand the logic behind "threshold".

**Comment by tolikzinovyev** [comment]
> Looks good, just one thing, if we choose `d: search_width`, to be consistent as @tolikzinovyev said, let's rename `t:search_index` (or `t:search_counter`)

`search_index` sounds good

**Comment by curiecrypt** [comment]
I used `search_counter` cause we used `retry_counter` for $v$. I can change if you wish

**Comment by curiecrypt** [comment]
> Looks good, I'm just not sure about `nf`. What are the tradeoffs between different options? Maybe I just don't understand the logic behind "threshold".

I think the `threshold` word better complies with the literature, but I am not insistent about it. I tried to choose the most voted-on options.

**Comment by tolikzinovyev** [comment]
> I used `search_counter` cause we used `retry_counter` for v . I can change if you wish

They definitely should be consistent! I have a slight preference for "index".

**Comment by tolikzinovyev** [comment]
> > Looks good, I'm just not sure about `nf`. What are the tradeoffs between different options? Maybe I just don't understand the logic behind "threshold".
> 
> I think the `threshold` word better complies with the literature, but I am not insistent about it. I tried to choose the most voted-on options.

Like from "threshold multi-signatures"?

**Comment by rrtoledo** [comment]
I think threshold comes more from the application (Peras) than Alba as there is no real link between threshold and lower bounds.
Would you prefer `lower_bound` for `nf`?

**Comment by curiecrypt** [comment]
> I think threshold comes more from the application (Peras) than Alba as there is no real link between threshold and lower bounds. Would you prefer `lower_bound` for `nf`?

`lower_bound` sounds good

**Comment by tolikzinovyev** [comment]
Any thoughts about "proven set size"? I feel like it says more than "lower bound". The prover proves that he has `nf` elements (`> nf` actually).

By the way, are we talking about the unweighted case only or both? If the later, is "size" the right word for us (applies to `np` too)?

**Comment by curiecrypt** [comment]
> Any thoughts about "proven set size"? I feel like it says more than "lower bound". The prover proves that he has `nf` elements (`> nf` actually).
> 
> By the way, are we talking about the unweighted case only or both? If the later, is "size" the right word for us (applies to `np` too)?

I think we should consider a generic name, for all versions.
However, `proven_set_size` does not cover the exact meaning in my head. Isn't it possible to have a proof that has a size `>nf`. In that case, this name does not make sense. Also, `lower_bound` means that we at least want a set of size `nf` but it could be larger. 

Actually, the terms and variable names are a bit confusing. But, the I deduce this when I read the paper. If I deduce it in a wrong way, then we must be so careful to explain it better to other users.

**Comment by rrtoledo** [comment]
I think @curiecrypt and I prefer putting the emphasis on the lower bound idea, hence `threshold` or `lower_bound`.

I think "set_size" might be misleading (it is not the set size but a lower bound on it |Sp|>nf) and "proven" does not bring much, just the idea of goal. 

I remember we mentioned in the past `target_bound`, that could be an adequate compromise.

**Comment by rrtoledo** [comment]
So far, this is only for the centralized PR.

When we implement (decentralised) weight and reuse the centralised code, I expect us to call telescope_centralised.Prove on a new set with specific parameter (from weighted/decentralised_weighted.setup). So the variables inside of centralised should not matter to (decentralised) weight too much.

**Comment by tolikzinovyev** [comment]
@curiecrypt @rrtoledo I think you misunderstood what I mean by "proven set size". It is **not** what the prover has, it is what the verifier is convinced the prover has. Specifically, the verifier knows the prover has `nf+1` elements, but he cannot know the prover has `nf+2` elements.

**Comment by rrtoledo** [comment]
We forgot a few parameters:
--- 
**lambda_sec**
- soundness_param
---
**lambda_rel**
- completeness_param
---
**h**
- round_hash
---
**h_usize**
- round_number

**Comment by rrtoledo** [comment]
> I think we should consider a generic name, for all versions. However, `proven_set_size` does not cover the exact meaning in my head. Isn't it possible to have a proof that has a size `>nf`. In that case, this name does not make sense. Also, `lower_bound` means that we at least want a set of size `nf` but it could be larger.
> 
> Actually, the terms and variable names are a bit confusing. But, the I deduce this when I read the paper. If I deduce it in a wrong way, then we must be so careful to explain it better to other users.

Technically you are right. If the `lambda`s are big enough, or `np/nf` small enough, `u` can be greater than `nf` indeed! Hence `u` can also be greater than `|Sp|`. I wonder however if you are not getting confused now between the "proof size" `u` and `nf`. haha

I agree we have to be very careful with the variable names. We do not want to confuse anyone, starting with us.

**Comment by tolikzinovyev** [comment]
So is `nf` vs `nf+1` the only drawback of "proven set size", or there are others? Just trying to understand pros and cons.

My strategy is this: if I know what alba does and I see a term like "proven set size", will I know what it refers to (e.g., np, nf, or something else)? With "proven set size", it most likely refers to nf. With "threshold", I really wouldn't know if it's np, nf or maybe something else. "lower bound" almost certainly refers to nf given that it's in "approximate lower bound arguments", but maybe it's a lower bound on the number of elements the proven needs to succeed. "target" seems too general (i.e. not very descriptive), I would certainly need to look up what it refers to.

Feel free to disagree with me. I just want to synchronize us on this.

**Comment by rrtoledo** [comment]
I want to dismiss any potential confusion. The name "proven set size" to me can be confusing and defintiely is not accurate for the reasons I listed before (nf/nf+1, and let's steer from any "prove" related word).

 > nf given that it's in "approximate lower bound arguments", but maybe it's a lower bound on the number of elements the proven needs to succeed.

It also is if you look at the definition of soundness.

The word "target" is general indeed, I only suggested it to replace "proven" and make sure to differentiate nf's name from the "dfs_bound". We could have "lower_bound", "set_bound", ...

**Comment by rrtoledo** [comment]
> !Future!
> I remember that I commented that we should comply with the names in the paper.
> However, limit does not seem to cover its actual functionality.

**Comment by rrtoledo** [comment]
> @curiecrypt @tolikzinovyev We forgot a few parameters and variable names:
> 
> **lambda_sec**
> 
> * soundness_param `R`
> 
> **lambda_rel**
> 
> * completeness_param `R`
> 
> **h (= hash(v,t, element_sequence))**
> 
> * round_hash `R`
> 
> **h_usize (= h1( v,t, element_sequence ) = Uni_np(h) )**
> 
> * round_number `R`
> * round_id `R`
> 
> **limit**
> 
> * limit `R`
> * step `R`

@curiecrypt @tolikzinovyev Make sure to vote here too!

**Comment by rrtoledo** [comment]
> what is the reason for calling the function parameter hash? Isn't that just an array of bytes?
For the oracles

**Comment by tolikzinovyev** [comment]
> let's steer from any "prove" related word

What's wrong with "proven" though? It refers to "proof of knowledge".

**Comment by tolikzinovyev** [comment]
> > @curiecrypt @tolikzinovyev We forgot a few parameters and variable names:
> > **lambda_sec**
> > 
> > * soundness_param `RG`
> > * lambda_sec `T`
> > 
> > **lambda_rel**
> > 
> > * completeness_param `RG`
> > * lambda_rel `T`
> > 
> > **h (= hash(v,t, element_sequence))**
> > 
> > * round_hash `RG`
> > 
> > **h_usize (= h1( v,t, element_sequence ) = Uni_np(h) )**
> > 
> > * round_number `R`
> > * round_id `RG`
> > 
> > **limit**
> > 
> > * limit `R`
> > * step `RG`
> 
> @curiecrypt @tolikzinovyev Make sure to vote here too!

We should try to keep variable names short also I think (soundness_param, completeness_param).

What does `limit` signify, number of steps made or number of steps remaining?

**Comment by tolikzinovyev** [comment]
How is `h` different from `H1`?
